numpy/core/src/multiarray/methods.c

```diff
@@ -2694,6 +2696,152 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+    int ndim = PyArray_NDIM(self);
+    npy_intp *strides = PyArray_STRIDES(self);
+    npy_intp *shape = PyArray_SHAPE(self);
+
+    for (int i = 0; i < ndim; ++i)
+    {
+        if (strides[i] % itemsize != 0) {
+            PyErr_SetString(PyExc_RuntimeError,
+                            "DLPack only supports strides which are a multiple of itemsize.");
+            return NULL;
+        }
+    }
+
+    DLDataType managed_dtype;
+    PyArray_Descr *dtype = PyArray_DESCR(self);
+    
+    managed_dtype.bits = 8 * itemsize;
+    managed_dtype.lanes = 1;
+    if (PyDataType_ISSIGNED(dtype))
+    {
+        managed_dtype.code = kDLInt;
+    }
+    else if (PyDataType_ISUNSIGNED(dtype))
+    {
+        managed_dtype.code = kDLUInt;
+    }
+    else if (PyDataType_ISFLOAT(dtype))
+    {
+        managed_dtype.code = kDLFloat;
+    }
+    else if (PyDataType_ISCOMPLEX(dtype))
+    {
+        managed_dtype.code = kDLComplex;
+    }
+    else
+    {
+        PyErr_SetString(PyExc_TypeError,
+                        "DLPack only supports signed/unsigned integers, float and complex dtypes.");
+        return NULL;
+    }
+
+    DLManagedTensor *managed = malloc(sizeof(DLManagedTensor));
+    if (managed == NULL)
+    {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Could not allocate the DLManagedTensor struct.");
+        return NULL;
+    }
+
+    managed->dl_tensor.data = PyArray_DATA(self);
+    managed->dl_tensor.device.device_type = kDLCPU;
+    managed->dl_tensor.device.device_id = 0;
+    managed->dl_tensor.dtype = managed_dtype;
+
+
+    int64_t *managed_shape = malloc(sizeof(int64_t) * ndim);
+    if (managed_shape == NULL)
+    {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Could not allocate the DLManagedTensor struct shape.");
+        free(managed);
+        return NULL;
+    }
+
+    int64_t *managed_strides = malloc(sizeof(int64_t) * ndim);
+    if (managed_strides == NULL)
+    {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Could not allocate the DLManagedTensor struct strides.");
+        free(managed);
+        free(managed_shape);
+        return NULL;
+    }
+
+    for (int i = 0; i < ndim; ++i)
+    {
+        managed_shape[i] = shape[i];
+        managed_strides[i] = strides[i];
+    }
+
+    managed->dl_tensor.ndim = ndim;
+    managed->dl_tensor.shape = managed_shape;
+    managed->dl_tensor.strides = managed_strides;
+    managed->dl_tensor.byte_offset = 0;
+    managed->manager_ctx = self;
+    managed->deleter = array_dlpack_deleter;
+
+    PyObject *capsule = PyCapsule_New(managed, NPY_DLPACK_CAPSULE_NAME, array_dlpack_capsule_deleter);
+    if (capsule != NULL)
+    {
+        PyArray_INCREF(self);
+    }
+    else
+    {
+        free(managed);
+        free(managed_shape);
+        free(managed_strides);
+    }
+    return capsule;
+}
+
+static PyObject *
+array_dlpack_device(PyArrayObject *self, PyObject *args)
```

Comment:
```suggestion
array_dlpack_device(PyArrayObject *NPY_UNUSED(self), PyObject *NPY_UNUSED(args))
```
although I am not sure the compilers actually complain...